pkg/stacks/truncate/truncate.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package truncate provides functions for truncating Kubernetes values in a
+// predictable way offering mildly collision safe values usable in
+// deterministic field searches.
+package truncate
+
+import (
+	// sha1 is not cryptographically secure, but that is not a goal. we require
+	// predictable and uniform text transformation. Any checksum function with
+	// even distribution would suffice.
+	"crypto/sha1" // nolint:blacklist
+	"encoding/base32"
+	"fmt"
+	"strings"
+)
+
+const (
+	// LabelNameLength is the max length of a name fragment in a label
+	LabelNameLength = 63
+
+	// LabelValueLength is the max length of a label value
+	LabelValueLength = 63
+
+	// ResourceNameLength is the max length of a resource name or a label with
+	// optional prefix and name
+	ResourceNameLength = 253
+
+	// DefaultSuffixLength is the length of combined separator and checksum
+	// characters that will be used as a truncation suffix in Label and Resource
+	// truncation functions
+	DefaultSuffixLength = 6
+
+	sha1Length = 40
+)
+
+// Truncate replaces the suffixLength number of trailing characters from str
+// with a consistent hash fragment based on that string. The suffix will include
+// a leading hyphen.
+// An error will be returned if the truncation length is less than the
+// suffixLength, or truncation length is greater than sha1Length, or the suffix
+// length is less than 2.
```

Comment:
How was this minimum chosen?

Reply:
Suffix length is an argument to `Truncate`. Each of the `Label*` and `ResourceName` helper functions have a predefined length that I pulled from the ether.  36^5 = 60,466,176.  While this may exceed (or not, I couldn't find a reference for it) the max resources per namespace, that is not the goal here.  The goal is to prevent name collisions on truncated resource names, "{really long namespace name}.stack-ms-print" should not be confused with "{same really long namespace name}.stack-ms-paint" when fetching host resources using a source resource as a reference.

any reference to "sha1" should be replaced with hash, digest, or sum.  I don't want to reflect that implementation choice in the comment.

minimum suffix length of 2 offers "-[a-z0-9]"